core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -114,20 +114,23 @@ class Partition(val topic: String,
   def isUnderReplicated: Boolean =
     isLeaderReplicaLocal && inSyncReplicas.size < assignedReplicas.size
 
-  def getOrCreateReplica(replicaId: Int = localBrokerId): Replica = {
-    assignedReplicaMap.getAndMaybePut(replicaId, {
+  def getOrCreateReplica(replicaId: Int = localBrokerId, isNew: Boolean = false): Replica = {
+    val replica = assignedReplicaMap.getAndMaybePut(replicaId, {
       if (isReplicaLocal(replicaId)) {
         val config = LogConfig.fromProps(logManager.defaultConfig.originals,
                                          AdminUtils.fetchEntityConfig(zkUtils, ConfigType.Topic, topic))
-        val log = logManager.createLog(topicPartition, config)
-        val checkpoint = replicaManager.highWatermarkCheckpoints(log.dir.getParentFile.getAbsolutePath)
+        val log = logManager.createLog(topicPartition, config, isNew)
+        val checkpoint = replicaManager.highWatermarkCheckpoints(log.dir.getParent)
         val offsetMap = checkpoint.read
         if (!offsetMap.contains(topicPartition))
           info(s"No checkpointed highwatermark is found for partition $topicPartition")
         val offset = math.min(offsetMap.getOrElse(topicPartition, 0L), log.logEndOffset)
         new Replica(replicaId, this, time, offset, Some(log))
       } else new Replica(replicaId, this, time)
     })
+    if (isReplicaLocal(replicaId) && !logManager.isLogOnline(topicPartition))
```

Comment:
We need this check to address the issue we have been discussing in the other threads, i.e. the partition may have been removed in `LogManager` but not in `ReplicaManager`. In this case, replicaId will be in the assignedReplicaMap even though the replica is offline. This additional check will make sure that `getOrCreateReplica()` will throw KafkaStorageException in this case.

More specifically, here is the currently workflow if LogCleaner encounters IOException:

- `LogManager.handleLogDirFailure()` will remove this log dir and partitions and inform controller via zookeeper.
- Controller sends LeaderAndIsrRequest for all partitions on this broker
- `Partition.getOrCreateReplica()` for partitions on the offline log directory will throw KafkaStorageException, which in turn triggers `ReplicaManager.handleLogDirFailure()` so that those partitions can be removed from `ReplicaManager` as well.